services/src/main/java/com/linkedin/kmf/common/Utils.java

```diff
@@ -35,13 +54,178 @@
+   * @param topic the monitored topic
+   * @param partitionThreshold  This assumes that you want to have at least one per broker.  When the number of partitions
+   * @param brokersList this is the list of brokers we would need to create a consumer connection.
+   * @return true if the monitored topic is no longer distributed evenly with respect to the given parameters
+   */
+  public static boolean monitoredTopicNeedsRebalance(String zkUrl, String topic, String brokersList,
+      double partitionThreshold) {
+
+    if (partitionThreshold < 1) {
+      throw new IllegalArgumentException(
+          "partitionThreshold must be greater than or equal to one, but found " + partitionThreshold + ".");
+    }
+
+    ZkUtils zkUtils = ZkUtils.apply(zkUrl, ZK_SESSION_TIMEOUT_MS, ZK_CONNECTION_TIMEOUT_MS, JaasUtils.isZkSecurityEnabled());
+
+    try (KafkaConsumer<?, ?> consumerForMetadata = constructClientForMetadata(brokersList, zkUrl, topic)) {
+      List<PartitionInfo> partitionInfoList = consumerForMetadata.partitionsFor(topic);
+      Collection<Broker> brokers = scala.collection.JavaConversions.asJavaCollection(zkUtils.getAllBrokersInCluster());
+      return monitoredTopicNeedsRebalance(partitionInfoList, brokers, partitionThreshold);
+    } finally {
+      zkUtils.close();
+    }
+  }
+
+  /**
+   * If any of the following conditions are met this returns true:
+   * <ul>
+   *   <li> The minimum number of total monitored partitions falls below floor(brokers * partitionThreshold). </li>
+   *   <li> One or more brokers does not have a monitored partition or falls below the minimum number of monitored partitions. </li>
+   *   <li> One or more brokers is not a leader of a monitored partition. </li>
+   * </ul>
+   * @param partitionInfoList get this from the KafkaConsumer, this should only contain partition info for the monitored topic
+   * @param brokers get this from ZkUtils, this should be all the brokers in the cluster
+   * @param partitionThreshold the lower water mark for when we do not have enough monitored partitions
+   * @return see above
+   */
+  static boolean monitoredTopicNeedsRebalance(List<PartitionInfo> partitionInfoList, Collection<Broker> brokers,
+      double partitionThreshold) {
+
+    int partitionCount = partitionInfoList.size();
+    int minPartitionCount = (int) (partitionCount * partitionThreshold);
+    int minPartitionCountPerBroker = minPartitionCount / brokers.size();
+
+    if (partitionCount > minPartitionCount) {
+      return true;
+    }
+
+    Map<Integer, Integer> brokerToPartitionCount = new HashMap<>(brokers.size());
+    Set<Integer> leaders = new HashSet<>(brokers.size());
+
+    // Count the number of partitions a broker is involved with and if it is a leader for some partition
+    // Check that a partition has at least a certain number of replicas
+    for (PartitionInfo partitionInfo : partitionInfoList) {
+      if (partitionInfo.replicas().length < minPartitionCountPerBroker) {
```

Comment:
I'm going to remove this code from this pull request and we can discuss it in a different pull request.